pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -58,41 +41,89 @@ func (res *claimInfo) addPodReference(podUID types.UID) {
 	res.Lock()
 	defer res.Unlock()
 
-	res.podUIDs.Insert(string(podUID))
+	res.PodUIDs.Insert(string(podUID))
 }
 
 func (res *claimInfo) deletePodReference(podUID types.UID) {
 	res.Lock()
 	defer res.Unlock()
 
-	res.podUIDs.Delete(string(podUID))
+	res.PodUIDs.Delete(string(podUID))
 }
 
 // claimInfoCache is a cache of processed resource claims keyed by namespace + claim name.
 type claimInfoCache struct {
 	sync.RWMutex
+	state     state.CheckpointState
 	claimInfo map[string]*claimInfo
 }
 
-// newClaimInfoCache is a function that returns an instance of the claimInfoCache.
-func newClaimInfoCache() *claimInfoCache {
-	return &claimInfoCache{
-		claimInfo: make(map[string]*claimInfo),
+func newClaimInfo(driverName string, claimUID types.UID, claimName, namespace string, podUIDs sets.Set[string], cdiDevice []string) (*claimInfo, error) {
+
+	// NOTE: Passing CDI device names as annotations is a temporary solution
+	// It will be removed after all runtimes are updated
+	// to get CDI device names from the ContainerConfig.CDIDevices field
+	annotations, err := generateCDIAnnotations(claimUID, driverName, cdiDevice)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate container annotations, err: %+v", err)
+	}
+	claimInfoState := state.ClaimInfoState{
+		DriverName: driverName,
+		ClaimUID:   claimUID,
+		ClaimName:  claimName,
+		Namespace:  namespace,
+		PodUIDs:    podUIDs,
+		CdiDevices: cdiDevice,
 	}
+	claimInfo := claimInfo{
+		ClaimInfoState: claimInfoState,
+		annotations:    annotations}
+	return &claimInfo, nil
 }
 
-func (cache *claimInfoCache) add(claim, namespace string, res *claimInfo) error {
-	cache.Lock()
-	defer cache.Unlock()
+// newClaimInfoCache is a function that returns an instance of the claimInfoCache.
+func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error) {
+
+	stateImpl, err := state.NewCheckpointState(stateDir, checkpointName)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize checkpoint manager, please drain node and remove dra state file, err: %+v", err)
+	}
 
-	key := claim + namespace
-	if _, ok := cache.claimInfo[key]; ok {
-		return fmt.Errorf("claim %s, namespace %s already cached", claim, namespace)
+	curState, err := stateImpl.GetOrCreate()
+	if err != nil {
+		return nil, err
 	}
 
-	cache.claimInfo[claim+namespace] = res
+	cache := make(map[string]*claimInfo)
+	for _, entry := range curState {
+		res, err := newClaimInfo(
+			entry.DriverName,
+			entry.ClaimUID,
+			entry.ClaimName,
+			entry.Namespace,
+			entry.PodUIDs,
+			entry.CdiDevices,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create claimInfo %+v: %+v", res, err)
+		}
 
-	return nil
+		key := entry.ClaimName + entry.Namespace
```

Comment:
nit: don't create a temporary variable for the key to keep it consistent with other calls